src/box/iproto.cc

```diff
@@ -1013,24 +1013,25 @@ iproto_msg_new(struct iproto_connection *con)
 }
 
 /**
- * Signal input unless it's blocked on I/O or stopped.
+ * Signal input unless it's blocked on I/O, stopped or is in replication.
  */
 static inline void
 iproto_connection_feed_input(struct iproto_connection *con)
 {
 	assert(con->state == IPROTO_CONNECTION_ALIVE);
-	if (!ev_is_active(&con->input) && rlist_empty(&con->in_stop_list))
+	if (!ev_is_active(&con->input) && rlist_empty(&con->in_stop_list) &&
+	    !con->is_in_replication)
 		ev_feed_event(con->loop, &con->input, EV_CUSTOM);
 }
 
 /**
- * Signal output unless it's blocked on I/O.
+ * Signal output unless it's blocked on I/O or is in replication.
  */
 static inline void
 iproto_connection_feed_output(struct iproto_connection *con)
 {
 	assert(con->state == IPROTO_CONNECTION_ALIVE);
-	if (!ev_is_active(&con->output))
+	if (!ev_is_active(&con->output) && !con->is_in_replication)
```

Comment:
Does it mean that after a replication request is enqueued in the iproto thread all the pending message responses will be lost?

Actually I think this is fine, but I still want to bring this to attention of other reviewers.

Why I think this is fine: a user should either wait for all the pending request responses prior to issuing a replication request in a single connection or have a separate connection for all the requests he wants to perform alongside the replication stream.

Reply:
Maybe, better be explicit about this and raise an error if a replication request is received while there are active requests in tx?

Reply:
Sounds good to me. Although it seems the existing `tx.requests_in_progress` metric wouldn't be enough for this.
We'd have to count already decoded but not yet processed requests on the iproto side.

Reply:
> Does it mean that after a replication request is enqueued in the iproto thread all the pending message responses will be lost?

No, it shouldn't. What do you call "pending message responses"? If you mean responses that were written to `obuf *iproto_connection::obuf`, but have not yet been written to the socket `iostream iproto_connection::io`, then they will not be lost anywhere, and will be written after the processing of the replication-request is completed.
If you mean requests (`iproto_msg`'s from "tx" thread) that are queued at `cbus_endpoint` "net{iproto thread->id}", then everything is fine with them too. The corresponding data has already been written to `obuf` in the "tx" thread. They will be processed in the "iproto" thread, they simply will not cause data to be written from `obuf` to the socket (`iproto_connection_feed_output` won't do anything). But this data will be written to the socket later after the processing of the replication-request is completed.

Reply:
In an good way, we should either cancel all pending requests **that the user has not yet read** or send them all without loss. But we will not be able to cancel all pending requests that the user has not yet read, because `obuf *iproto_connection:::obuf` is flushed to the socket asynchronously. That is, at the moment when we parse the replication request in "iproto" thread, some random number of responses could be written to the socket, and the user does not know how much he needs to subtract before starting to read the response to the replication request. Moreover, it may turn out that the last request that we tried to write to the socket was partially written. And if we right now become the owner of the socket in the “tx” thread right now and write a response to a replication request to it, we will get invalid msgpuck. Therefore, in my opinion, the best solution would be to either force it to wait for all previous requests to be written to the socket, or make it the user's responsibility (the user is required to wait for responses to all previous requests before sending a replication request).

Reply:
Ok, I see, thanks for the explanation.

>  If you mean responses that were written to obuf *iproto_connection::obuf, but have not yet been written to the socket iostream iproto_connection::io

Yes, this is what I meant. Given your explanation I still think the responces are "lost" in a sense that a client first receives a responce to his replication request and only then receives a responce to his previous request (BTW, I'm not sure it's possible to continue normal iproto operation after replication is being processed).

I think we should return an error then if we receive a replication request before we have sent responces to all the previous user's requests.

Another option would be to pause replication request processing (in iproto thread) until all responces to previous requests are written.

@locker, which option do you prefer?

Reply:
> I think we should return an error then if we receive a replication request before we have sent responces to all the previous user's requests.
> 
> Another option would be to pause replication request processing (in iproto thread) until all responces to previous requests are written.

I prefer the first option (rejecting replication request) because it should be easier to implement AFAICS. Nobody should send a replication request before processing all pending responses.

Reply:
I have implemented it, please have a look.